_drafts/gdb-python-api-pypi-packages.md

```diff
@@ -0,0 +1,453 @@
+from the post mentioned above to make use of a third party package installed
+through PyPi.
+
+<!-- excerpt end -->
+
+In the post [Automate Debugging with GDB Python
+API]({% post_url 2019-07-02-automate-debugging-with-gdb-python-api %}), Chris
+worked through an example Python GDB command which printed UUIDs stored in a
+Linked List. In this post, we want to expand upon this `uuid_list_dump` command
+and make the output more readable for humans using the
+[PTable](https://pypi.org/project/PTable/) package from PyPi. We will print the
+contents in an ASCII formatted table.
+
+**Before:**
+
+```
+(gdb) uuid_list_dump s_list_head
+Args Passed: s_list_head
+Found a Linked List with 10 nodes:
+0: Addr: 0x200070c0, uuid: 9d5d4440-8b5a-235e-a9ef-ec3f919c7a41
+1: Addr: 0x200070a8, uuid: 8afe211b-197e-e475-8ec6-878558dd24e6
+2: Addr: 0x20007088, uuid: 0f84cd49-d683-4217-e7a3-399373c6e6f1
+3: Addr: 0x20007070, uuid: 23b05b1b-a32a-8288-4419-0e30f26ba3ae
+4: Addr: 0x20007050, uuid: 3ca835e5-8f2a-cdc2-40c1-cf52716a7229
+5: Addr: 0x20007038, uuid: 30d4349d-3a0d-0fbd-2fa1-f70fd968f4d9
+6: Addr: 0x20007018, uuid: 437bbe90-1689-9d7e-77c6-2f269888f5b4
+7: Addr: 0x20007000, uuid: 5e20fb38-a84e-a614-9325-562444df598d
+8: Addr: 0x20006fe0, uuid: 7a96092c-a557-7464-c4af-1528a4e957db
+9: Addr: 0x20006fc8, uuid: 2dcf4629-04b4-78d8-68a7-ff3f2bf1fcd9
+```
+
+**After:**
+
+```
+(gdb) uuid_list_dump s_list_head
+Args Passed: s_list_head
+Found a Linked List with 10 nodes:
++-----+-------------+--------------------------------------+
+| Idx |     Addr    |                 UUID                 |
++-----+-------------+--------------------------------------+
+|  0  | 0x1002002d0 | 008e7897-970e-3462-07d7-5e47a158298e |
+|  1  | 0x1002002b0 | deb2f6b5-e24c-4fdd-9f88-67bdce1ff261 |
+|  2  | 0x100200290 | 818fc33a-1e0f-c02c-4455-7ac8ab50c9b2 |
+|  3  | 0x100200270 | 0ce10698-6d61-ff34-a11e-19fd3650e9b7 |
+|  4  | 0x100200250 | d3feeda5-d5f3-d9e4-5bfa-6cc351e220ae |
+|  5  | 0x100200230 | e9c5f1b0-c415-8ae5-9b4d-39f6f7e8a105 |
+|  6  | 0x100200210 | 02353981-84b8-14a2-9cb4-5a672acae548 |
+|  7  | 0x1002001f0 | 48336241-f30d-23e5-5f30-d1c8ed610c4b |
+|  8  | 0x100200080 | 17119854-2f11-2d05-58f5-6bd688079992 |
+|  9  | 0x100200060 | a7f1d92a-82c8-d8fe-434d-98558ce2b347 |
++-----+-------------+--------------------------------------+
+```
+
+This is incredibly easy to do using the PTable package, as using the package is
+as simple as:
+
+```
+x = PrettyTable()
+x.field_names = ["Field Name 1", ...]
+x.add_row(["Value 1", ...])
+...
+print(x)
+```
+
+Lets go ahead and try it out!
+
+_Like Interrupt? [Subscribe](http://eepurl.com/gpRedv) to get our latest
+posts straight to your mailbox_
+
+## Setting up GDB's Python and PyPi
+
+If you want to know the solution right now and skip over our investigation step,
+continue from [here](#setting-syspath-within-gdbinit).
+
+### Initial Notes
+
+Importing PyPi packages within your current installation might work already, but
```

Comment:
I would remove these three paragraphs, I think they mostly add confusion.

Reply:
Yeah....they aren't good paragraphs. I didn't have them at first but then realized I'm sure people have this working for them just because they got lucky, or because they don't realize they are editing their global python.

Reply:
Added a section about virtualenv and posted a link to my favorite tutorial.